proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java

```diff
@@ -107,8 +107,10 @@ public void handle(PacketWrapper packet) throws Exception
     @Override
     public void handle(KeepAlive alive) throws Exception
     {
-        server.setSentPingId( alive.getRandomId() );
-        con.setSentPingTime( System.currentTimeMillis() );
+        if (server.getSentKeepAlives().getFirst().getMillis() + bungee.getConfig().getTimeout() < System.currentTimeMillis()) {
+            throw new IllegalStateException("Did not recieve keepalive in time from " + con.getName());
```

Comment:
Any clever trick to move this exception over to UpstreamBridge so proper exception handling is done? Atm this would connect you to fallback but one should get disconnected instead.

Reply:
@md-5 ?

Reply:
Short of running it in the generic packet handler method, probably not.
You could move it there, but it would need to be optimized (cache last one )